rust-symcrypt/src/hmac.rs

```diff
@@ -1,180 +1,318 @@
+/// doing so would lead to use-after-free and inconsistent states.
 impl HmacSha256State {
-    pub fn new(key: &[u8]) -> Self {
+    pub fn new(key: &[u8]) -> Result<Self, SymCryptError> {
         let mut instance = HmacSha256State {
-            state: symcrypt_sys::SYMCRYPT_HMAC_SHA256_STATE::default(),
-            expanded_key: symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default(),
+            inner: Box::pin(HmacSha256Inner { 
+                state: symcrypt_sys::SYMCRYPT_HMAC_SHA256_STATE::default(),
+                expanded_key: symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default() 
+            }),
         };
-        unsafe {
-            symcrypt_sys::SymCryptHmacSha256ExpandKey(
-                &mut instance.expanded_key,
+        unsafe { // SAFETY: FFI calls
+            match symcrypt_sys::SymCryptHmacSha256ExpandKey(
+                &mut instance.inner.expanded_key,
                 key.as_ptr(),
                 key.len() as symcrypt_sys::SIZE_T,
-            );
-            symcrypt_sys::SymCryptHmacSha256Init(&mut instance.state, &instance.expanded_key);
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => {
+                    symcrypt_sys::SymCryptHmacSha256Init(&mut instance.inner.state, &instance.inner.expanded_key);
+                    Ok(instance)
+                },
+                err => Err(err.into())
+            }
         }
-        instance
     }
+}
 
-    pub fn append(&mut self, data: &[u8]) {
-        unsafe {
+impl Hmac for HmacSha256State {
+    type Result = [u8; SHA256_HMAC_RESULT_SIZE];
+
+    fn append(&mut self, data: &[u8]) {
+        unsafe { // SAFETY: FFI calls
             symcrypt_sys::SymCryptHmacSha256Append(
-                &mut self.state,
+                &mut self.inner.state,
                 data.as_ptr(),
                 data.len() as symcrypt_sys::SIZE_T,
             )
         }
     }
 
-    pub fn result(&mut self, result: &mut [u8; SHA256_HMAC_RESULT_SIZE]) {
-        unsafe {
-            symcrypt_sys::SymCryptHmacSha256Result(&mut self.state, result.as_mut_ptr());
+    fn result(mut self) -> Self::Result {
+        let mut result = [0u8; SHA256_HMAC_RESULT_SIZE];
+        unsafe { // SAFETY: FFI calls
+            symcrypt_sys::SymCryptHmacSha256Result(&mut self.inner.state, result.as_mut_ptr());
+        }
+        result
+    }
+    
+    fn copy(&self) -> Box<Self> {
+        let mut new_state = HmacSha256State {
+            inner: Box::pin(HmacSha256Inner { 
+                state: symcrypt_sys::SYMCRYPT_HMAC_SHA256_STATE::default(),
+                expanded_key: symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default() 
+            }),
+        };
+        unsafe { // SAFETY: FFI calls
+            symcrypt_sys::SymCryptHmacSha256StateCopy(&self.inner.state, &self.inner.expanded_key, &mut new_state.inner.state);
         }
+        Box::new(new_state)
     }
```

Comment:
This won't work. The copy points back to the original expanded key.

Will need to do on your Expand Key object instead.